From 0b5dfa1898b173791d89199992ea60a1dc1cb954 Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Fri, 5 Sep 2014 02:16:02 -0400
Subject: [pci] Add function 0 DMA alias quirk for Ricoh devices

Message-id: <20140905021602.32574.65583.stgit@gir.stowe>
Patchwork-id: 91363
O-Subject: [RHEL7.1 PATCH 04/17] PCI: Add function 0 DMA alias quirk for Ricoh devices
Bugzilla: 1136537
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Alex Williamson <alex.williamson@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1136537
Upstream Status: ec637fb2d4f7a7167c323dc1d22837d1270a21bc

Signed-off-by: Jarod Wilson <jarod@redhat.com>

Conflicts:
  ./drivers/pci/quirks.c
    Hunk #1; the reference to 'dma_alias_devfn' had to be adjusted due
    to KABI (see RHEL7's back-port related to upstream commit 31c2b81).

commit ec637fb2d4f7a7167c323dc1d22837d1270a21bc
Author: Alex Williamson <alex.williamson@redhat.com>
Date:   Thu May 22 17:07:49 2014 -0600

    PCI: Add function 0 DMA alias quirk for Ricoh devices

    The existing quirk for these devices (pci_get_dma_source()) doesn't really
    solve the problem; re-implement it using the DMA alias iterator.  We'll
    come back later and remove the existing quirk and dma_source interface.
    Note that device ID 0xe822 is typically function 0 and 0xe230 has been
    tested to not need the quirk and are therefore removed versus the
    equivalent dma_source quirk.  If there exist in other configurations we can
    re-add them.

    Link: https://bugzilla.redhat.com/show_bug.cgi?id=605888
    Tested-by: Pat Erley <pat-lkml@erley.org>
    Signed-off-by: Alex Williamson <alex.williamson@redhat.com>
    Signed-off-by: Bjorn Helgaas <bhelgaas@google.com>
---

diff --git a/drivers/pci/quirks.c b/drivers/pci/quirks.c
index 71cb4cd..1508802 100644
--- a/drivers/pci/quirks.c
+++ b/drivers/pci/quirks.c
@@ -3337,6 +3337,23 @@ int pci_dev_specific_reset(struct pci_dev *dev, int probe)
  return -ENOTTY;
 }
 
+static void quirk_dma_func0_alias(struct pci_dev *dev)
+{
+ if (PCI_FUNC(dev->devfn) != 0) {
+  dev->pci_dev_rh->dma_alias_devfn =
+   PCI_DEVFN(PCI_SLOT(dev->devfn), 0);
+  dev->dev_flags |= PCI_DEV_FLAGS_DMA_ALIAS_DEVFN;
+ }
+}
+
+/*
+ * https://bugzilla.redhat.com/show_bug.cgi?id=605888
+ *
+ * Some Ricoh devices use function 0 as the PCIe requester ID for DMA.
+ */
+DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_RICOH, 0xe832, quirk_dma_func0_alias);
+DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_RICOH, 0xe476, quirk_dma_func0_alias);
+
 static struct pci_dev *pci_func_0_dma_source(struct pci_dev *dev)
 {
  if (!PCI_FUNC(dev->devfn))
-- 
1.7.1